### PR TITLE
Functional Blob and Bucket classes

### PR DESCRIPTION
**Pull request by mziccard**
This PR adds support for functional `Blob` and `Bucket`, as discussed in #55. First two commits are borrowed by @aozarov. Overall the PR does:
1.  Split `ListResult` into interface (`ListResult`) and implementation (`BaseListResult`)
2.  Add `BlobListResult` for paginated lists of `Blob` (to be used by `Bucket`)
3.  Add tests for `BlobListResult`
4.  Add `Bucket` and `Blob` classes
5.  Add tests for `Blob` and `Bucket`

Commit d8fa7bb does 1, 2 and 3 while commit 44cefb0 does 4 and 5.

### CODE REVIEW

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/ListResult.java [hunk 1]
```diff
@@ -16,67 +16,28 @@
...
-    this.pageFetcher = pageFetcher;
-    this.cursor = cursor;
-    this.results = results;
-  }
+public interface ListResult<T> extends Iterable<T> {
 
   /**
    * Returns the cursor for the nextPage or {@code null} if no more results.
+   * 
+   * @return the string cursor for next page
```
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/ListResult.java]: I think the preferred style in this case (simple "Returns XXX" statement) would be without the extra "@return".
This [link](https://google.github.io/styleguide/javaguide.html#s7.2-summary-fragment) does directly state that, but is related.

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/ListResult.java [hunk 2]
```diff
@@ -16,67 +16,28 @@
...
    */
-  public String nextPageCursor() {
-    return cursor;
-  }
+  public String nextPageCursor();
```
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/ListResult.java]: No need for the explicit "public" method qualifier for interface elements.

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+
+  /**
+   * Copy this blob.
+   *
+   * @param target target blob
+   * @param options source blob options
+   * @return the copied blob
+   * @throws StorageException upon failure
+   */
+  public Blob copyTo(BlobInfo target, BlobSourceOption... options) {
```
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: I wonder if we want this method (and the one bellow). 
GCS#copy makes it clear that source is only bucket/name. I am afraid that people may think that this method
does more than that (e.g. getting this blobs content-type if one not provided by target).

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/BlobListResult.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Implementation of a paginated list of Google Cloud storage {@code Blob}.
+ */
+public class BlobListResult implements ListResult<Blob> {
+
+  private final ListResult<BlobInfo> infoList;
+  private final transient Storage storage;
```
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/BlobListResult.java]: transient is not need (for now) as this class is not Serializable.

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/BlobListResult.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  private transient List<Blob> results;
+
+  public BlobListResult(Storage storage, ListResult<BlobInfo> infoList) {
+    this.storage = storage;
```
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/BlobListResult.java]: checkNotNull

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/BlobListResult.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    ListResult<BlobInfo> nextPageInfoList = infoList.nextPage();
+    if (nextPageInfoList == null) {
+      return null;
+    }
+    return new BlobListResult(storage, nextPageInfoList);
+  }
+
+  @Override
+  public Iterator<Blob> iterator() {
+    if (results == null) {
```
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/BlobListResult.java]: Alternative, you can use Guava [Iterators#transform](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/collect/Iterators.html#transform(java.util.Iterator,%20com.google.common.base.Function)) and I am not sure we need to keep results as a member variable.

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java (hunk 3 above)
Comment (mziccard) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: Could we keep these methods and find a way to document that this blob metadata are not used to integrate target's?

Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: Can you check what other veneer libraries do in this case?
Yes, if we keep it we need to at least make the javadoc very clear but I would also
think how to simplify the second version (which looks hairier compare to the other methods in this class).

Comment (mziccard) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: -  **gcloud-go** has a single function `copyObject` that works similarly to our `Datastore.copy`
-  **gcloud-node** has a `File` object with a `copy(target)` method where target can be either the bucket name as string (object name is preserved), a `Bucket` object (object name is preserved) or a `File` object.
-  **gcloud-python** has a `Bucket` object with a `copy_blob` method that takes a source `Blob` object, a destination `Bucket` object and a new name as string
-  **gcloud-ruby** has a `File` object with a `copy` method that takes destination bucket and destination name as strings

I would keep the methods and state:
> The current `Blob` object is only used to retrieve bucket and blob name. This method does not update remote object's metadata other than `name` and `bucket`.

Why do you thing that the second version needs to be simplified? It takes one more parameter wrt the previous that is needed only if you need to set some `BlobTargetOption`. Would you prefer having both parameters wrapped in a `CopyOptions` object?

Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: As for **gcloud-go** I assume you mean "our `Blob.copy`"?
Looks like most libraries takes destination as a string.. Do they copy their own (this) metadata
to the destination?
 
> This This method does not update remote object's metadata ...

But it does, using the values from target `BlobInfo`, just not from the source.
We can iterate on the Javadoc to make it more clear however I think it is easier
to add later than remove later.

I agree, having this functionality in the other libraries makes it more compelling to have it
here as well, but maybe we can start with something simpler such as.
copyTo(bucket, sourceOptions...) and copyTo(bucket, name, sourceOptions....)
and copy the metadata by default?

> Why do you thing that the second version needs to be simplified? 

I consider the options as an advanced feature that most user will not care about and
would rather make it as transparent as possible.  Also, a side note, I am not sure the param
sequence feels right (target, sourceOptions, targetOptions)..

Comment (mziccard) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: Sorry for the tons of typos. For **gcloud-go** I meant that they have something similar to our `Storage.copy`.  
> Do they copy their own (this) metadata to the destination?

**gcloud-node**, **gcloud-ruby** and **gcloud-python** seem not to pass any metadata. **gcloud-go** has a metadata parameter that if specified is used to override metadata.

> But it does, using the values from target BlobInfo, just not from the source.

You are right.

> maybe we can start with something simpler such as.
copyTo(bucket, sourceOptions...) and copyTo(bucket, name, sourceOptions....)
and copy the metadata by default?

Absolutely. If we get rid of the `target` Blob parameter we can use current `Blob`'s info to possibly update metadata. Can we go with this?

> I consider the options as an advanced feature that most user will not care about and
would rather make it as transparent as possible. 

Agree. But that method is only there to allow a user to set both options, which as you said is not the common case. Normal users have their `copy(target)` right?

> Also, a side note, I am not sure the param sequence feels right (target, sourceOptions, targetOptions)..

Not sure any other permutation would make more sense :)

**Comment by mziccard** [comment]
I added an iterator class just for clarity as we might be putting our hands again on `BlobListResult`. However, if you prefer `transform` because of consistency with the rest of the library I can switch to it!

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+    this.info = BlobInfo.of(checkNotNull(bucket), checkNotNull(blob));
+  }
+
+  /**
+   * Return the blob's information.
+   */
+  public BlobInfo info() {
+    return info;
+  }
+
```
Comment (ajkannan) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: Could this ever return stale info (and if so, could this counter intuitive for the user)?  An alternative could be to get a fresh BlobInfo each time, but that would probably be an expensive operation.

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/Bucket.java [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.gcloud.storage.Storage.BlobSourceOption;
+import com.google.gcloud.storage.Storage.BlobTargetOption;
+import com.google.gcloud.storage.Storage.BucketSourceOption;
+import com.google.gcloud.storage.Storage.BucketTargetOption;
+import java.util.LinkedList;
+
```
Comment (ajkannan) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Bucket.java]: (very nit): I think the empty line should go before the linked list import.

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java (hunk 7 above)
Comment (mziccard) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: Actually, yes it can return stale info. I had a look at what the others do:
-  **gcloud-python**: return a stale copy
-  **gcloud-ruby**: return a stale copy but provide a method to fetch most recent metadata: `file.reload`
-  **gcloud-node**: have a getter that always fetches an updated version

What do you think? We might have `info()` that returns stale information and add a `reload()` function that fetches metadata and updates local `Blob`'s info. Maybe @aozarov can give us some thoughts on this:)

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java (hunk 3 above)
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: > Absolutely. If we get rid of the target Blob parameter we can use current Blob's info to possibly update metadata. Can we go with this?

Sounds good to me.

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java (hunk 7 above)
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: I could have sworn that we had a reload/refresh method... at some point I had it (could even be a change on my laptop that I never pushed).

As for **gcloud-node**,  I wonder. if, lets say, bucket returns files (from list, getAlll, ...) doesn't it
already get the metadata for them?

We could add a `reload`. We could also say that Blob is immutable and pinned to specific info and update operations (and `reload`) returns a new/updated Blob. 

As oppose to what **gcloud-node** does I think it make more sense to *pin* Blob
to its info (especially when we want to execute *update_or_delete-if-was-not-change*, using the options).

We could also optimize (with more work) and add support for [selected fields](https://developers.google.com/resources/api-libraries/documentation/storage/v1/java/latest/com/google/api/services/storage/StorageRequest.html#setFields(java.lang.String)). Maybe using options.
With such support we can make operations like list only returns names and change Blob
to accept either BlobInfo or bucket/name and for the latter lazily load the complete info upon request.

What do you think?

Comment (mziccard) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: > As oppose to what gcloud-node does I think it make more sense to pin Blob
to its info (especially when we want to execute update_or_delete-if-was-not-change, using the options).

If we pin `Blob` to its info wouldn't it make sense to always set the `x-goog-if-metageneration-match` header? Otherwise we might have situations like:
```java
Blob oldBlob = new Blob(storage, oldInfo);
Blob newBlob = oldBlob.update(newInfo);
Blob newerBlob = oldBlob.update(newerInfo); // still works
```
Having a `Blob` pinned to its info but using only its `name` and `bucket` to issue requests might be a bit counter-intuitive.

> We could also optimize (with more work) and add support for selected fields. Maybe using options. With such support we can make operations like list only returns names and change Blob to accept either BlobInfo or bucket/name and for the latter lazily load the complete info upon request.

This sounds cool. But I would not put this as high priority task (nor include it in this PR) as optimization is not the main focus of the library. I would open an issue for this.

Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: > Having a Blob pinned to its info but using only its name and bucket to issue requests might be a bit counter-intuitive.

I am not sure about that. I mean, the user has that option by passing BlobSourceOption.generationMatch and/or the BlobSourceOptions.metagenerationMatch
options. Though clearly if we choose to go that route documentation should make it clear
that operations are not enforcing the version check by default.

> This sounds cool. But I would not put this as high priority task (nor include it in this PR) as optimization is not the main focus of the library. I would open an issue for this.

Yes,  I also didn't think we should do it in this PR but I do think having an optimized code (if it does not compromise the API) is something that we should care for. Also, for listing of many
files this could have a significant benefit.

Comment (ajkannan) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: I vote we go with returning stale info and having the reload function.  I think pinning a Blob to it's info would be fine, and I'd vote for forcing the version match by default to discourage situations like the `oldBlob`, `newBlob`, `newestBlob` Marco brought up.  I think pinning a Blob to it's info would probably be a little bit clearer for users because they're less likely to view it as a mutable, up-to-date copy of what's stored in GCS and write buggy code due to version issues.

I don't know if this is relevant/correct, but by pinning a version, it seems like the Blob class would just provide syntactic sugar.  Was that the original point, or was there additional functionality that we lose by pinning a Blob to it's BlobInfo?

Comment (mziccard) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: It is also true that the user might want to write something like:
```java
BlobInfo info = BlobInfo.of("bucket", "name");
Blob blob = new Blob(storage, info);
blob.update(info.toBuilder().contentType("application/json").build());
```
In these cases adding by default the `metagenerationMatch` options makes no sense.
Commits up to 2e90926 go the @aozarov way: immutable Blob and Bucket and no generation match by default. I'll add some detailed javadocs to the update methods then maybe you can check whether this deserves merging :)

> Was that the original point, or was there additional functionality that we lose by pinning a Blob to it's BlobInfo?

Yes I think that `Blob` and `Bucket` are meant to be syntactic sugar and I can not thing of any functionality loss we could have by making them immutable.

Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: > In these cases adding by default the metagenerationMatch options makes no sense.
Commits up to 2e90926 go the @aozarov way.

I don't think that would be a problem as that info does not have a value for `metageneration`
and therefore will not enforce it. Are you suggesting that it might confuse the caller?
I am actually on the fence on this one.

Options are:
- enforce metageneration check and have only an option to disable this check.
- don't enforce metageneration and keep all options, letting the user choose.

Either way we should be immutable and make the documentation clear.
I want to say that I actually like the first option (as I feel it is slightly more intuitive for blob)
however I also think that it is probably less desired for bucket.

@mziccard pick any option you prefer and lets create a separate issue for it and wait
until we get some more opinions.

Comment (mziccard) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: As the commit are now there's no default `metageneration` check and that choice is documented:

> By default no checks are made on the metadata generation of the current blob. If you want to update the information only if the current blob metadata are at their latest version use the `metagenerationMatch` option: 
`blob.update(newInfo, blobTargetOption.metagenerationMatch());`.

We can merge as is (if you don't see other issues) and open a new issue.

Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: SGTM

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+   * @param options blob delete options
+   * @return true if blob was deleted
+   * @throws StorageException upon failure
+   */
+  public boolean delete(BlobSourceOption... options) {
+    return storage.delete(info.bucket(), info.name(), convert(info, options));
+  }
+
+  /**
+   * Copy this blob to the target bucket, preserving its name. Possibly update metadata.
```
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java]: Possibly update metadata -> possibly copying also some of the metadata (e.g. content-type)?

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/Bucket.java [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+   * 
+   * @param blobNames names of the requested blobs
+   * @throws StorageException upon failure
+   */
+  public List<Blob> getAll(String... blobNames) {
+    BatchRequest.Builder batch = BatchRequest.builder();
+    for (String blobName : blobNames) {
+      batch.get(info.name(), blobName);
+    }
+    List<Blob> blobs = new LinkedList<>();
```
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/Bucket.java]: Nit, could be new ArrayList<>(blobNames.length);

File: gcloud-java-storage/src/main/java/com/google/gcloud/storage/ListResult.java [hunk 11]
```diff
@@ -16,67 +16,19 @@
...
-  public ListResult(NextPageFetcher<T> pageFetcher, String cursor, Iterable<T> results) {
-    this.pageFetcher = pageFetcher;
-    this.cursor = cursor;
-    this.results = results;
-  }
+public interface ListResult<T> extends Iterable<T> {
 
   /**
-   * Returns the cursor for the nextPage or {@code null} if no more results.
+   * Return the cursor for the nextPage or {@code null} if no more results.
```
Comment (aozarov) [inline gcloud-java-storage/src/main/java/com/google/gcloud/storage/ListResult.java]: Why did you change Returns to Return? Here and in other places. I think the prefer style is [Returns](https://google.github.io/styleguide/javaguide.html#s7.2-summary-fragment).

**Comment by aozarov** [comment]
@mziccard looks good. see 3 last comments above and I think we are ready to merge :+1:

**Comment by mziccard** [comment]
@aozarov fixed :)

**Comment by aozarov** [comment]
Great! Thanks.
